src/nvim/lib/rbuffer.h

```diff
@@ -0,0 +1,285 @@
+// Ring buffer implementation. This is basically an array that wraps read/write
+// pointers around the memory region. It should be more efficient than the old
+// RBuffer which required memmove() calls to relocate read/write positions.
+//
+// The main purpose of RBuffer is simplify memory management when reading from
+// uv_stream_t instances:
+//
+// - The event loop writes data to a RBuffer, advancing the write pointer
+// - The main loop reads data, advancing the read pointer
+// - If the buffer becomes full(size == capacity) the rstream is temporarily
+//   stopped(automatic backpressure handling)
+//
+// Reference: http://en.wikipedia.org/wiki/Circular_buffer
+#ifndef NVIM_LIB_RBUFFER_H
+#define NVIM_LIB_RBUFFER_H
+
+#include <assert.h>
+#include <stddef.h>
+#include <string.h>
+
+#include "nvim/memory.h"
+#include "nvim/vim.h"
+
+
+// Macros that simplify working with the read/write pointers directly by hiding
+// ring buffer wrap logic. Some examples:
+//
+// - Pass the write pointer to a function(`write_data`) that incrementally
+//   produces data, returning the number of bytes actually written:
+//
+//       RBUFFER_EACH_WRITE_PTR(rbuf, ptr, cnt)
+//         rbuffer_produced(rbuf, write_data(state, ptr, cnt));
+//
+// - Pass the read pointer to a function(`read_data`) that incrementally
+//   consumes data, return the number of bytes actually read:
+//
+//       RBUFFER_EACH_READ_PTR(rbuf, ptr, cnt)
+//         rbuffer_consumed(rbuf, read_data(state, ptr, cnt));
```

Comment:
I forgot to update the usage of these macros which have different semantics. These ones are only useful iterate over the all the data and only consume later:

```c
size_t consumed = 0;
RBUFFER_EACH_READ_PTR(rbuf, ptr, cnt)
  if ((consumed += read_data(state, ptr, cnt)) < cnt) break;
rbuffer_consumed(rbuf, consumed);
```